### PR TITLE
Decoupling allocation from writing in HTTP/2 outbound flow control

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

The current DefaultHttp2RemoteFlowController's writePendingBytes currently operates in 2 passes. The first allocates bytes and optionally writes some frames. The second pass just loops across all active streams and writes all remaining bytes.

If streams can be removed/added as a side effect of writing (EOS or error) then we need to take more care when the write actually occurs. Moving all of the writes to the second loop (across active streams) is simpler since we can just make a copy of the list and not worry about any restructuring of the priority tree that may result.

Modifications:

Modified DefaultHttp2RemoteFlowController.writePendingBytes to only allocate bytes on the first pass and then write any allocated bytes on the second pass.

Result:

Side effects resulting from writing should no longer impact the flow control algorithm.

### CODE REVIEW

**Comment by nmittler** [comment]
@Scottmitch with the coming changes related to stream removal, I suspect this is going to be important.  Can you take a look?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 1]
```diff
@@ -232,21 +232,27 @@ private void writePendingBytes() {
...
-            writeChildren(connectionStream, connectionWindow);
-            for (Http2Stream stream : connection.activeStreams()) {
-                writeChildNode(state(stream));
+            // Allocate the bytes for the connection window to the streams, but do not write.
+            allocateBytesForTree(connectionStream, connectionWindow);
+
+            // Now write all of the allocated bytes. Copying the activeStreams array to avoid
+            // side effects due to stream removal/addition which might occur as a result
+            // of end-of-stream or errors.
+            for (Http2Stream stream : connection.activeStreams().toArray(new Http2Stream[0])) {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: I wonder if we can avoid this copy operation. Also it seems like other scenarios where we loop over the active streams may have similar issues.  I may take a stab at https://github.com/netty/netty/pull/3537#discussion_r27249115 and see if it makes sense.

**Comment by Scottmitch** [comment]
@nmittler - I'm getting a compile error:

> /Users/scott_mitchell/git/netty/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java:325: two or more consecutive empty lines

**Comment by Scottmitch** [comment]
Also `allocated()` it not used anymore.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 1 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Possibly, but I think that should be a separate PR.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Depending on the ratio of how many streams get bytes allocated vs how many active stream there are it may be beneficial to maintain a set member variable to track which streams have bytes allocated (or need to write) for this pass....I think I was previously passing a set as an argument though the recursive calls...and this may have been before I change the structure to what it is now....

**Comment by nmittler** [comment]
Weird, I wasn't getting that build error. Anyway, I removed the line ... PTAL.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 1 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: @Scottmitch that sounds reasonable.  I've filed #3550 for this.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: +1 to both your comments.  I've got a PR pending to avoid copy.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 2]
```diff
@@ -365,13 +361,23 @@ void window(int initialWindowSize) {
          */
         void allocate(int bytes) {
             allocated += bytes;
+            // Also artificially reduce the streamable bytes for this tree to give the appearance
+            // that the data has been written. This will be restored when we write the allocated bytes.
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: `This will be restored when we write the allocated bytes.` - > `This will be restored before the allocated bytes are actually written.`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 3]
```diff
@@ -232,21 +233,28 @@ private void writePendingBytes() {
...
+                state(stream).writeAllocatedBytes();
             }
             flush();
         }
     }
 
     /**
-     * Write the children of {@code parent} in the priority tree. This will allocate bytes by stream weight.
-     * @param parent The parent of the nodes which will be written.
+     * This will allocate bytes by stream weight for the entire tree rooted at {@code parent}, but does not write
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Shouldn't we mention priority in here too? Something like...

```This will allocate bytes according to stream weight and priority...`

**Comment by Scottmitch** [comment]
@nmittler - Great work!  Just a few nits.  Feel free to cherry-pick once resolved.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 2 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 3 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Done.

**Comment by nmittler** [comment]
Thanks for the quick review! Cherry-picked into master and 4.1.
